core/cas-server-core-services-registry/src/main/java/org/apereo/cas/services/AbstractServicesManager.java

```diff
@@ -155,48 +160,134 @@ public RegisteredService findServiceBy(final Service service) {
 
     @Override
     public RegisteredService findServiceBy(final long id) {
-        val result = this.services.get(id);
+        val result = this.services.get(id, k -> this.serviceRegistry.findServiceById(id));
         return validateRegisteredService(result);
     }
 
+    @Override
+    public <T extends RegisteredService> T findServiceBy(final long id, final Class<T> clazz) {
+        val result = this.serviceRegistry.findServiceById(id, clazz);
+        if (result != null) {
+            services.get(result.getId(), k -> result);
+        }
+        return result;
+    }
+
+    @Override
+    public RegisteredService findServiceByName(final String name) {
+        if (StringUtils.isBlank(name)) {
+            return null;
+        }
+
+        var service = services.asMap().values().stream()
+                .filter(r -> r.getName().equals(name))
+                .findFirst()
+                .orElse(null);
+
+        if (service == null) {
+            LOGGER.trace("The service with name {} is not found in the cache, try to find it from [{}]",
+                    name, serviceRegistry.getName());
+            service = serviceRegistry.findServiceByExactServiceName(name);
+            if (service != null) {
+                services.put(service.getId(), service);
+                LOGGER.trace("The service is found in [{}] and populated to the cache {}  ", serviceRegistry.getName(),
```

Comment:
```suggestion
                LOGGER.trace("The service is found in [{}] and populated to the cache [{}]", serviceRegistry.getName(),
```

Also, formatting of parameters is off here. Should be: "The service [{}] is found in service registry [{}]" and cached.